### ISSUE TITLE
Different behaviors when setting edata attribute?

### ISSUE THREAD

**Issue by davidsonic** [dmlc/dgl#721]
## ❓ Questions and Help

I was trying out an experiment with GCN. One with the line setting edata (as below) and one without. 
![image](https://user-images.githubusercontent.com/15846113/61581555-a3d5d780-aad4-11e9-9897-f30a2105f19d.png)

I didn't use this edge attribute in the code. However, the training result is different. With edata set:
![image](https://user-images.githubusercontent.com/15846113/61581538-738e3900-aad4-11e9-8ce3-07a03e371295.png)

Without:
![image](https://user-images.githubusercontent.com/15846113/61581567-e3042880-aad4-11e9-8d4b-38c4f2c78570.png)

It seems that edata attribute have some effects on training? BTW, the GCN implementation i'm using is: https://github.com/dmlc/dgl/blob/master/examples/pytorch/gcn/gcn_mp.py

**Comment by mufeili** [comment]
Could you please try fixing random seed and see if the difference is still significant? Note that you need multiple lines to get deterministic behavior in PyTorch:

```python
import numpy as np
import random
import torch
import torch.backends.cudnn as cudnn

random.seed(1)
numpy.random.seed(1)
torch.manual_seed(1)

# And if this is on gpu
torch.cuda.manual_seed(1)
cudnn.benchmark = False
cudnn.deterministic = True
```

**Comment by davidsonic** [comment]
Thanks for your response, I did have seed fixed similar to your reply.
![image](https://user-images.githubusercontent.com/15846113/61582916-ca503e80-aae5-11e9-9cb0-d31eb3a3fd4e.png)
I still observe the difference though. 
I'll add cudnn lines and try some more.

**Comment by VoVAllen** [comment]
Our kernel includes `atomicAdd` operation on gpu when calling `update_all(fn.copy_u(...), fn.sum(...))`, which is not deterministic, and may not be able to fix easily.

Is it deterministic on cpu? @ylfdq1118

**Comment by yzh119** [comment]
@VoVAllen atomicAdd/Max is only used in CUDA.
btw, atomic operations would be deprecated later.

**Comment by jermainewang** [comment]
That variance looks huge. Could you paste the result of ten different runs?

**Comment by davidsonic** [comment]
I haven't notices such huge variances ever since. Close it for now.
